// firebase-function/lib/example/AddPersonFunction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddPersonFunction = void 0;
const src_1 = require("../src");
class AddPersonFunction {
    auth;
    databaseReference;
    logger;
    parameters;
    constructor(parameterContainer, auth, databaseReference, logger) {
        this.auth = auth;
        this.databaseReference = databaseReference;
        this.logger = logger;
        this.logger.log('AddPersonFunction.constructor', { auth: auth }, 'notice');
        const parameterParser = new src_1.ParameterParser({
            id: new src_1.ParameterBuilder('string', src_1.Guid.fromString),
            name: new src_1.ValueParameterBuilder('string'),
            age: new src_1.ValueParameterBuilder('number')
        }, this.logger.nextIndent);
        parameterParser.parse(parameterContainer);
        this.parameters = parameterParser.parameters;
    }
    async execute() {
        this.logger.log('AddPersonFunction.executeFunction', {}, 'info');
        await this.databaseReference.child('persons').child(this.parameters.id.guidString).set({
            name: this.parameters.name,
            age: this.parameters.age
        }, 'encrypt');
    }
}
exports.AddPersonFunction = AddPersonFunction;
