// firebase-function/lib/example/GetAnimalsFunction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetAnimalsFunction = void 0;
const src_1 = require("../src");
class GetAnimalsFunction {
    auth;
    databaseReference;
    logger;
    parameters;
    constructor(parameterContainer, auth, databaseReference, logger) {
        this.auth = auth;
        this.databaseReference = databaseReference;
        this.logger = logger;
        this.logger.log('GetAnimalsFunction.constructor', { auth: auth }, 'notice');
        const parameterParser = new src_1.ParameterParser({}, this.logger.nextIndent);
        parameterParser.parse(parameterContainer);
        this.parameters = parameterParser.parameters;
    }
    async execute() {
        this.logger.log('GetAnimalsFunction.executeFunction', {}, 'info');
        const snapshot = await this.databaseReference.child('animals').snapshot();
        if (!snapshot.exists)
            return [];
        return snapshot.compactMap(snapshot => {
            if (snapshot.key === null)
                return null;
            return {
                id: snapshot.key,
                name: snapshot.child('name').value()
            };
        });
    }
}
exports.GetAnimalsFunction = GetAnimalsFunction;
