// firebase-function/lib/example/GetPersonRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetPersonRequest = void 0;
const src_1 = require("../src");
class GetPersonRequest {
    databaseReference;
    logger;
    parameters;
    constructor(parameterContainer, databaseReference, logger) {
        this.databaseReference = databaseReference;
        this.logger = logger;
        this.logger.log('GetPersonRequest.constructor', undefined, 'notice');
        const parameterParser = new src_1.ParameterParser({
            id: new src_1.ParameterBuilder('string', src_1.Guid.fromString)
        }, this.logger.nextIndent);
        parameterParser.parse(parameterContainer);
        this.parameters = parameterParser.parameters;
    }
    async execute() {
        this.logger.log('GetPersonRequest.executeFunction', {}, 'info');
        const snapshot = await this.databaseReference.child('persons').child(this.parameters.id.guidString).snapshot();
        if (!snapshot.exists)
            return null;
        return {
            ...snapshot.value('decrypt'),
            id: this.parameters.id.guidString
        };
    }
}
exports.GetPersonRequest = GetPersonRequest;
