// firebase-function/lib/example/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
const admin = __importStar(require("firebase-admin"));
const firebaseFunctions_1 = require("./firebaseFunctions");
const src_1 = require("../src");
const testUtils_1 = require("../src/testUtils");
const privateKeys = {
    cryptionKeys: {
        encryptionKey: new src_1.FixedLength(Uint8Array.from([]), 32),
        initialisationVector: new src_1.FixedLength(Uint8Array.from([]), 16),
        vernamKey: new src_1.FixedLength(Uint8Array.from([]), 32)
    },
    callSecretKey: 'abc',
    databaseUrl: 'url/to/my/database'
};
admin.initializeApp();
// eslint-disable-next-line @typescript-eslint/no-unused-vars
async function test() {
    const firebaseApp = new testUtils_1.FirebaseApp({}, privateKeys.cryptionKeys, privateKeys.callSecretKey, {
        functionsRegion: 'europe-west1',
        databaseUrl: privateKeys.databaseUrl
    });
    await firebaseApp.functions.function('person').function('add').call({ id: 'some-guid', name: 'Max Mustermann', age: 25 });
    await firebaseApp.functions.function('person').function('get').request({ id: 'some-guid' });
    await firebaseApp.functions.function('animal').function('get').call({});
}
module.exports = (0, src_1.createFirebaseFunctions)(() => privateKeys, {}, firebaseFunctions_1.firebaseFunctions);
