// firebase-function/lib/test/MockDatabase.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chai_1 = require("chai");
const testUtils_1 = require("../src/testUtils");
describe('MockDatabase', () => {
    it('test', async () => {
        const databaseScheme = testUtils_1.MockDatabaseReference.createDatabaseScheme(encrypt => ({
            value1: 12,
            value2: 'asdf',
            value3: encrypt(false),
            value4: {
                value5: 'a',
                value6: 'b',
                value7: 'c'
            }
        }));
        const databaseReference = new testUtils_1.MockDatabaseReference(databaseScheme);
        (0, chai_1.expect)((await databaseReference.snapshot()).child('value1').value()).to.be.equal(12);
        (0, chai_1.expect)((await databaseReference.child('value2').snapshot()).value()).to.be.equal('asdf');
        (0, chai_1.expect)((await databaseReference.snapshot()).child('value3').value('decrypt')).to.be.equal(false);
        (0, chai_1.expect)((await databaseReference.child('value4').snapshot()).reduce('', (r, s) => r + s.value())).to.be.equal('abc');
        await databaseReference.child('value1').set(5);
        await databaseReference.child('value3').set(true, 'encrypt');
        await databaseReference.child('value2').remove();
        (0, chai_1.expect)((await databaseReference.snapshot()).child('value1').value()).to.be.equal(5);
        (0, chai_1.expect)((await databaseReference.snapshot()).child('value3').value('decrypt')).to.be.equal(true);
        (0, chai_1.expect)((await databaseReference.child('value2').snapshot()).exists).to.be.false;
        await databaseReference.child('value4').set({ value5: 'x', value6: 'y', value7: 'z' });
        (0, chai_1.expect)((await databaseReference.child('value4').snapshot()).reduce('', (r, s) => r + s.value())).to.be.equal('xyz');
        await databaseReference.remove();
        (0, chai_1.expect)((await databaseReference.snapshot()).exists).to.be.false;
    });
});
