// firebase-function/lib/test/ParameterContainer.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chai_1 = require("chai");
const crypter_1 = require("../src/crypter");
const logger_1 = require("../src/logger");
const src_1 = require("../src");
const parameter_1 = require("../src/parameter");
describe('Parameter container', () => {
    const cryptionKeys = {
        encryptionKey: new crypter_1.FixedLength(Uint8Array.from([0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0E, 0x0F, 0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19, 0x1A, 0x1B, 0x1C, 0x1D, 0x1E, 0x1F]), 32),
        initialisationVector: new crypter_1.FixedLength(Uint8Array.from([0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0E, 0x0F]), 16),
        vernamKey: new crypter_1.FixedLength(Uint8Array.from([0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0E, 0x0F, 0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19, 0x1A, 0x1B, 0x1C, 0x1D, 0x1E, 0x1F]), 32)
    };
    const crypter = new crypter_1.Crypter(cryptionKeys);
    const logger = logger_1.Logger.start(new logger_1.VerboseType('coloredVerbose'), 'paramter container test');
    it('parameters invalid', () => {
        (0, chai_1.expect)(() => new parameter_1.ParameterContainer({
            databaseType: new src_1.DatabaseType('testing')
        }, crypter, logger.nextIndent)).to.throw();
        (0, chai_1.expect)(() => new parameter_1.ParameterContainer({
            databaseType: new src_1.DatabaseType('testing'),
            parameters: 0
        }, crypter, logger.nextIndent)).to.throw();
        const parameterContainer = new parameter_1.ParameterContainer({
            databaseType: new src_1.DatabaseType('testing'),
            parameters: crypter.encodeEncrypt(null)
        }, crypter, logger.nextIndent);
        (0, chai_1.expect)(() => parameterContainer.parameter('value0', new parameter_1.ValueParameterBuilder('number'))).to.throw();
    });
    it('get invalid parameter', () => {
        const parameterContainer = new parameter_1.ParameterContainer({
            databaseType: new src_1.DatabaseType('testing'),
            parameters: crypter.encodeEncrypt({
                value1: undefined,
                value2: null,
                value3: 'asdf'
            })
        }, crypter, logger.nextIndent);
        (0, chai_1.expect)(() => parameterContainer.parameter('value0', new parameter_1.ValueParameterBuilder('number'))).to.throw();
        (0, chai_1.expect)(() => parameterContainer.parameter('value1', new parameter_1.ValueParameterBuilder('number'))).to.throw();
        (0, chai_1.expect)(() => parameterContainer.parameter('value2', new parameter_1.ValueParameterBuilder('number'))).to.throw();
        (0, chai_1.expect)(() => parameterContainer.parameter('value3', new parameter_1.ValueParameterBuilder('number'))).to.throw();
    });
    it('get parameter', () => {
        const parameterContainer = new parameter_1.ParameterContainer({
            databaseType: new src_1.DatabaseType('testing'),
            parameters: crypter.encodeEncrypt({
                value0: undefined,
                value1: true,
                value2: 'asdf',
                value3: 12,
                value5: { v: 'asdf' },
                value6: null
            })
        }, crypter, logger.nextIndent);
        (0, chai_1.expect)(parameterContainer.parameter('value0', new parameter_1.ValueParameterBuilder('undefined'))).to.be.undefined;
        (0, chai_1.expect)(parameterContainer.parameter('value1', new parameter_1.ValueParameterBuilder('boolean'))).to.be.true;
        (0, chai_1.expect)(parameterContainer.parameter('value2', new parameter_1.ValueParameterBuilder('string'))).to.be.equal('asdf');
        (0, chai_1.expect)(parameterContainer.parameter('value3', new parameter_1.ValueParameterBuilder('number'))).to.be.equal(12);
        (0, chai_1.expect)(parameterContainer.parameter('value5', new parameter_1.ValueParameterBuilder('object'))).to.be.deep.equal({ v: 'asdf' });
        (0, chai_1.expect)(parameterContainer.parameter('value6', new parameter_1.ValueParameterBuilder('object'))).to.be.null;
    });
    it('get parameter not crypted', () => {
        const parameterContainer = new parameter_1.ParameterContainer({
            databaseType: new src_1.DatabaseType('testing'),
            value0: undefined,
            value1: true,
            value2: 'asdf',
            value3: 12,
            value5: { v: 'asdf' },
            value6: null
        }, null, logger.nextIndent);
        (0, chai_1.expect)(parameterContainer.parameter('value0', new parameter_1.ValueParameterBuilder('undefined'))).to.be.undefined;
        (0, chai_1.expect)(parameterContainer.parameter('value1', new parameter_1.ValueParameterBuilder('boolean'))).to.be.true;
        (0, chai_1.expect)(parameterContainer.parameter('value2', new parameter_1.ValueParameterBuilder('string'))).to.be.equal('asdf');
        (0, chai_1.expect)(parameterContainer.parameter('value3', new parameter_1.ValueParameterBuilder('number'))).to.be.equal(12);
        (0, chai_1.expect)(parameterContainer.parameter('value5', new parameter_1.ValueParameterBuilder('object'))).to.be.deep.equal({ v: 'asdf' });
        (0, chai_1.expect)(parameterContainer.parameter('value6', new parameter_1.ValueParameterBuilder('object'))).to.be.null;
    });
});
