// firebase-function/lib/src/types/Result.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Result = void 0;
var Result;
(function (Result) {
    class Success {
        value;
        state = 'success';
        constructor(value) {
            this.value = value;
        }
        get error() {
            return null;
        }
        get valueOrError() {
            return this.value;
        }
        get() {
            return this.value;
        }
        map(mapper) {
            return new Result.Success(mapper(this.value));
        }
        mapError() {
            return this;
        }
    }
    Result.Success = Success;
    class Failure {
        error;
        state = 'failure';
        constructor(error) {
            this.error = error;
        }
        get value() {
            return null;
        }
        get valueOrError() {
            return this.error;
        }
        get() {
            // eslint-disable-next-line @typescript-eslint/no-throw-literal
            throw this.error;
        }
        map() {
            return this;
        }
        mapError(mapper) {
            return new Result.Failure(mapper(this.error));
        }
    }
    Result.Failure = Failure;
    function success(value) {
        return new Result.Success(value);
    }
    Result.success = success;
    function failure(error) {
        return new Result.Failure(error);
    }
    Result.failure = failure;
    function isSuccess(result) {
        return result.state === 'success';
    }
    Result.isSuccess = isSuccess;
    function isFailure(result) {
        return result.state === 'failure';
    }
    Result.isFailure = isFailure;
})(Result = exports.Result || (exports.Result = {}));
